webapi/CopilotChat/Controllers/ChatController.cs

```diff
@@ -183,16 +211,47 @@ private async Task RegisterPlannerSkillsAsync(CopilotChatPlanner planner, OpenAp
         }
 
         // Microsoft Graph
-        if (!string.IsNullOrWhiteSpace(openApiSkillsAuthHeaders.GraphAuthentication))
+        if (openApiSkillsAuthHeaders.TryGetValue("GRAPH", out string? GraphAuthHeader))
         {
             this._logger.LogInformation("Enabling Microsoft Graph skill(s).");
-            BearerAuthenticationProvider authenticationProvider = new(() => Task.FromResult(openApiSkillsAuthHeaders.GraphAuthentication));
+            BearerAuthenticationProvider authenticationProvider = new(() => Task.FromResult(GraphAuthHeader));
             GraphServiceClient graphServiceClient = this.CreateGraphServiceClient(authenticationProvider.AuthenticateRequestAsync);
 
             planner.Kernel.ImportSkill(new TaskListSkill(new MicrosoftToDoConnector(graphServiceClient)), "todo");
             planner.Kernel.ImportSkill(new CalendarSkill(new OutlookCalendarConnector(graphServiceClient)), "calendar");
             planner.Kernel.ImportSkill(new EmailSkill(new OutlookMailConnector(graphServiceClient)), "email");
         }
+
+        if (variables.TryGetValue("customPlugins", out string? customPluginsString))
+        {
+            CustomPlugin[]? customPlugins = JsonSerializer.Deserialize<CustomPlugin[]>(customPluginsString);
+
+            if (customPlugins != null)
+            {
+                foreach (CustomPlugin plugin in customPlugins)
+                {
+                    if (openApiSkillsAuthHeaders.TryGetValue(plugin.AuthHeaderTag.ToUpperInvariant(), out string? PluginAuthValue))
+                    {
+                        // Register the ChatGPT plugin with the planner's kernel.
+                        this._logger.LogInformation("Enabling {0} plugin.", plugin.NameForHuman);
+
+                        UriBuilder uriBuilder = new(plugin.ManifestDomain);
+                        // Expected manifest path as defined by OpenAI: https://platform.openai.com/docs/plugins/getting-started/plugin-manifest
+                        uriBuilder.Path = "/.well-known/ai-plugin.json";
+
+                        // TODO: Support other forms of auth. Currently, we only support user PAT or no auth.
+                        var requiresAuth = !plugin.AuthType.Equals("none", StringComparison.OrdinalIgnoreCase);
+                        BearerAuthenticationProvider authenticationProvider = new(() => Task.FromResult(PluginAuthValue));
+
+                        await planner.Kernel.ImportChatGptPluginSkillFromUrlAsync($"{plugin.NameForModel}Plugin", uriBuilder.Uri,
```

Comment:
nit:
```suggestion
                        await planner.Kernel.ImportChatGptPluginSkillFromUrlAsync(
                            $"{plugin.NameForModel}Plugin", 
                            uriBuilder.Uri,
```

Reply:
the formatter should have a preference here

Reply:
formatted